src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns_Constant.cs

```diff
@@ -0,0 +1,159 @@
+                case SpecialType.System_Int64: comparison = BinaryOperatorKind.LongEqual; break;
+                case SpecialType.System_UInt64: comparison = BinaryOperatorKind.ULongEqual; break;
+                default: throw ExceptionUtilities.UnexpectedValue(comparisonType.Name);
+            }
+
+            var equalsTest =
+                _factory.Binary(comparison, _factory.SpecialType(SpecialType.System_Boolean),
+                    _factory.Convert(comparisonType, _factory.Local(temp)),
+                    _factory.Convert(comparisonType, boundConstant));
+            return _factory.Sequence(temp, _factory.LogicalAnd(call, equalsTest));
+        }
+
+        private MethodSymbol SelectIntegralPatternHelper(BoundExpression boundConstant)
+        {
+            Debug.Assert(boundConstant.ConstantValue != null);
+            var constantValue = boundConstant.ConstantValue;
+            var node = boundConstant.Syntax;
+
+            // Depending on the *value* of the constant value, we select one of three helper functions.
+            // If it is within the range of values of an int, we use AsIntValue. If it is
+            // outside that range on the negative side, we use AsLargeNegative. If it is outside that range
+            // on the positive side, we use AsLargePositive.
+            Debug.Assert(constantValue.IsIntegral);
+            switch (constantValue.Discriminator)
+            {
+                case ConstantValueTypeDiscriminator.Int64:
+                    {
+                        // this can be inside the range of ints, or outside the range at either end, either too high or too low.
+                        long value = constantValue.Int64Value;
+                        return (value < int.MinValue) ? AsLargeNegativeHelper(node) :
+                               (value > int.MaxValue) ? AsLargePositiveHelper(node) :  AsIntValueHelper(node);
+                    }
+                case ConstantValueTypeDiscriminator.UInt32:
+                case ConstantValueTypeDiscriminator.UInt64:
+                    {
+                        // this can be inside the range of ints or outside the range at the positive side.
+                        ulong value = constantValue.UInt64Value;
+                        return (value > int.MaxValue) ? AsLargePositiveHelper(node) : AsIntValueHelper(node);
+                    }
+                default:
+                    {
+                        // all of the remaining integral types have value sets totally contained in the values of int.
+                        return AsIntValueHelper(node);
+                    }
+            }
+        }
+
+        private MethodSymbol AsLargeNegativeHelper(CSharpSyntaxNode node)
+        {
+            return MakePrivateImplementationHelper(
+                node,
+                PrivateImplementationDetails.AsLargeNegativeName,
+                (m, pi) => new SynthesizedAsLargeNegativeMethod(m, pi));
```

Comment:
Consider `module` instead of `m`. 
For `pi` we may be stuck (no good understandable and short name?).